gpcontrib/arenadata_toolkit/arenadata_toolkit--1.4--1.5.sql

```diff
@@ -0,0 +1,127 @@
+/* gpcontrib/arenadata_toolkit/arenadata_toolkit--1.4--1.5.sql */
+
+/*
+ This is part of arenadata_toolkit API for ADB Bundle.
+ This function collects information in db_files_current and db_files_history tables.
+ */
+CREATE OR REPLACE FUNCTION arenadata_toolkit.adb_collect_table_stats()
+RETURNS VOID
+AS $$
+DECLARE
+	ops oid[] DEFAULT (SELECT array_agg(oid) FROM pg_opclass WHERE opcname = 'timestamp_ops');
+BEGIN
+	IF NOT EXISTS (
+		SELECT
+		FROM pg_partition pp
+		JOIN pg_class cl1 ON pp.parrelid = cl1.oid
+		JOIN pg_partition_rule pr ON pr.paroid = pp.oid
+		JOIN pg_class cl2 ON cl2.oid = pr.parchildrelid
+		JOIN pg_namespace n1 ON cl1.relnamespace = n1.oid
+		JOIN pg_namespace n2 ON cl2.relnamespace = n2.oid
+		WHERE
+			CASE
+				WHEN pp.parclass[0] = ANY (ops)
+				THEN now() BETWEEN
+					CAST(substring(pg_get_expr(pr.parrangestart, pr.parchildrelid) FROM '#"%#"::%' FOR '#')
+						AS TIMESTAMP WITHOUT TIME ZONE)
+					AND
+					CAST(substring(pg_get_expr(pr.parrangeend, pr.parchildrelid) FROM '#"%#"::%' FOR '#')
+						AS TIMESTAMP WITHOUT TIME ZONE)
+				ELSE FALSE
+			END

```

Comment:
could you compare this solution with
```sql
			(pp.parclass[0] != ALL (ops) OR
			now() BETWEEN CAST(
				substring(pg_get_expr(pr.parrangestart, pr.parchildrelid) FROM '#"%#"::%' FOR '#')
				AS TIMESTAMP WITHOUT TIME ZONE)
			AND
			CAST(substring(pg_get_expr(pr.parrangeend, pr.parchildrelid) FROM '#"%#"::%' FOR '#')
				AS TIMESTAMP WITHOUT TIME ZONE)
			)
```
?

Reply:
what about plans?

Reply:
Here are plans if the logic is wrapped into a single query (to allow plan analysis by explain):

<details>
<summary> log </summary>


```
postgres=# 
postgres=# explain analyze SELECT
                FROM pg_partition pp
                JOIN pg_class cl1 ON pp.parrelid = cl1.oid
                JOIN pg_partition_rule pr ON pr.paroid = pp.oid
                JOIN pg_class cl2 ON cl2.oid = pr.parchildrelid
                JOIN pg_namespace n1 ON cl1.relnamespace = n1.oid
                JOIN pg_namespace n2 ON cl2.relnamespace = n2.oid
                WHERE
                        CASE
                                WHEN pp.parclass[0] = ANY(SELECT oid FROM pg_opclass WHERE opcname = 'timestamp_ops')
                                THEN now() BETWEEN
                                        CAST(substring(pg_get_expr(pr.parrangestart, pr.parchildrelid) FROM '#"%#"::%' FOR '#')
                                                AS TIMESTAMP WITHOUT TIME ZONE)
                                        AND
                                        CAST(substring(pg_get_expr(pr.parrangeend, pr.parchildrelid) FROM '#"%#"::%' FOR '#')
                                                AS TIMESTAMP WITHOUT TIME ZONE)
                                ELSE FALSE
                        END
                        AND
                        cl1.relname = 'db_files_history'
                        AND
                        n1.nspname = 'arenadata_toolkit'
                        AND
                        n2.nspname = 'arenadata_toolkit'
                        AND
                        pp.paristemplate = false
                        AND
                        pp.parkind = 'r'  --range
                        AND
                        pr.parrangestartincl = 't';
                                                                                                                                                                                                                                 QUERY 
PLAN                                                                                                                                                                                                                                  
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Hash Join  (cost=225.84..387.60 rows=4 width=0) (actual time=7.735..9.234 rows=1 loops=1)
   Hash Cond: (cl2.relnamespace = n2.oid)
   Extra Text: Hash chain length 1.0 avg, 1 max, using 1 of 131072 buckets.
   ->  Hash Join  (cost=224.74..386.45 rows=4 width=4) (actual time=7.620..8.979 rows=1 loops=1)
         Hash Cond: (cl2.oid = pr.parchildrelid)
         Extra Text: Hash chain length 1.0 avg, 1 max, using 1 of 131072 buckets.
         ->  Seq Scan on pg_class cl2  (cost=0.00..141.31 rows=5431 width=8) (actual time=0.007..0.696 rows=3474 loops=1)
         ->  Hash  (cost=224.70..224.70 rows=2 width=4) (actual time=7.475..7.475 rows=1 loops=1)
               Buckets: 131072  Batches: 1  Memory Usage: 1kB
               ->  Hash Join  (cost=69.66..224.70 rows=4 width=4) (actual time=6.605..7.472 rows=1 loops=1)
                     Hash Cond: (cl1.relnamespace = n1.oid)
                     Extra Text: Hash chain length 1.0 avg, 1 max, using 1 of 131072 buckets.
                     ->  Hash Join  (cost=68.56..223.56 rows=4 width=8) (actual time=6.497..7.227 rows=1 loops=1)
                           Hash Cond: (pp.oid = pr.paroid)
                           Join Filter: CASE WHEN (hashed SubPlan 1) THEN (('2024-04-15 14:49:29.723938+10'::timestamp with time zone >= ("substring"(pg_get_expr(pr.parrangestart, pr.parchildrelid), '#"%#"::%'::text, '#'::text))::t
imestamp without time zone) AND ('2024-04-15 14:49:29.723938+10'::timestamp with time zone <= ("substring"(pg_get_expr(pr.parrangeend, pr.parchildrelid), '#"%#"::%'::text, '#'::text))::timestamp without time zone)) ELSE false END
                           Extra Text: Hash chain length 2.0 avg, 6 max, using 991 of 32768 buckets.
                           ->  Hash Join  (cost=7.01..161.94 rows=4 width=40) (actual time=0.463..1.063 rows=1 loops=1)
                                 Hash Cond: (cl1.oid = pp.parrelid)
                                 Extra Text: Hash chain length 1.0 avg, 2 max, using 992 of 65536 buckets.
                                 ->  Seq Scan on pg_class cl1  (cost=0.00..154.89 rows=1 width=8) (actual time=0.010..0.505 rows=1 loops=1)
                                       Filter: (relname = 'db_files_history'::name)
                                       Rows Removed by Filter: 3473
                                 ->  Hash  (cost=7.00..7.00 rows=1 width=40) (actual time=0.382..0.382 rows=1001 loops=1)
                                       Buckets: 65536  Batches: 1  Memory Usage: 63kB
                                       ->  Seq Scan on pg_partition pp  (cost=0.00..7.00 rows=1 width=40) (actual time=0.008..0.273 rows=1001 loops=1)
                                             Filter: ((NOT paristemplate) AND (parkind = 'r'::"char"))
                           ->  Hash  (cost=58.00..58.00 rows=1 width=72) (actual time=1.436..1.436 rows=2001 loops=1)
                                 Buckets: 32768  Batches: 1  Memory Usage: 712kB
                                 ->  Seq Scan on pg_partition_rule pr  (cost=0.00..58.00 rows=1 width=72) (actual time=0.002..0.818 rows=2001 loops=1)
                                       Filter: parrangestartincl
                                       Rows Removed by Filter: 1
                           SubPlan 1
                             ->  Materialize  (cost=0.00..3.54 rows=4 width=4) (actual time=0.033..0.033 rows=3 loops=1)
                                   ->  Seq Scan on pg_opclass  (cost=0.00..3.53 rows=4 width=4) (actual time=0.012..0.028 rows=3 loops=1)
                                         Filter: (opcname = 'timestamp_ops'::name)
                                         Rows Removed by Filter: 199
                     ->  Hash  (cost=1.09..1.09 rows=1 width=4) (actual time=0.005..0.005 rows=1 loops=1)
                           Buckets: 131072  Batches: 1  Memory Usage: 1kB
                           ->  Seq Scan on pg_namespace n1  (cost=0.00..1.09 rows=1 width=4) (actual time=0.004..0.004 rows=1 loops=1)
                                 Filter: (nspname = 'arenadata_toolkit'::name)
                                 Rows Removed by Filter: 7
   ->  Hash  (cost=1.09..1.09 rows=1 width=4) (actual time=0.017..0.017 rows=1 loops=1)
         Buckets: 131072  Batches: 1  Memory Usage: 1kB
         ->  Seq Scan on pg_namespace n2  (cost=0.00..1.09 rows=1 width=4) (actual time=0.010..0.010 rows=1 loops=1)
               Filter: (nspname = 'arenadata_toolkit'::name)
               Rows Removed by Filter: 7
 Planning time: 5.166 ms
   (slice0)    Executor memory: 16713K bytes.  Work_mem: 712K bytes max.
 Memory used:  128000kB
 Optimizer: Postgres query optimizer
 Execution time: 9.416 ms
(51 rows)

postgres=# 
postgres=# 
postgres=# 
postgres=# 
postgres=# 
postgres=# 
postgres=# 
postgres=# 
postgres=# 
postgres=# 
postgres=# 
postgres=# 
postgres=# 
postgres=# explain analyze SELECT
                FROM pg_partition pp
                JOIN pg_class cl1 ON pp.parrelid = cl1.oid
                JOIN pg_partition_rule pr ON pr.paroid = pp.oid
                JOIN pg_class cl2 ON cl2.oid = pr.parchildrelid
                JOIN pg_namespace n1 ON cl1.relnamespace = n1.oid
                JOIN pg_namespace n2 ON cl2.relnamespace = n2.oid
                WHERE
                        pp.paristemplate = false
                        AND
                        n1.nspname = 'arenadata_toolkit'
                        AND
                        n2.nspname = 'arenadata_toolkit'
                        AND
                        cl1.relname = 'db_files_history'
                        AND
                        (pp.parclass[0] != ALL (SELECT oid FROM pg_opclass WHERE opcname = 'timestamp_ops') OR
                        now() BETWEEN CAST(
                                substring(pg_get_expr(pr.parrangestart, pr.parchildrelid) FROM '#"%#"::%' FOR '#')
                                AS TIMESTAMP WITHOUT TIME ZONE)
                        AND
                        CAST(substring(pg_get_expr(pr.parrangeend, pr.parchildrelid) FROM '#"%#"::%' FOR '#')
                                AS TIMESTAMP WITHOUT TIME ZONE)
                        )
                        AND
                        pp.parkind = 'r'  --range
                        AND
                        pr.parrangestartincl = 't';
                                                                                                                                                                                                                 QUERY PLAN            
                                                                                                                                                                                                      
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Hash Join  (cost=224.08..385.84 rows=4 width=0) (actual time=7.366..12.424 rows=1 loops=1)
   Hash Cond: (cl2.relnamespace = n2.oid)
   Extra Text: Hash chain length 1.0 avg, 1 max, using 1 of 131072 buckets.
   ->  Hash Join  (cost=222.98..384.69 rows=4 width=4) (actual time=7.253..12.125 rows=1 loops=1)
         Hash Cond: (cl2.oid = pr.parchildrelid)
         Extra Text: Hash chain length 1.0 avg, 1 max, using 1 of 131072 buckets.
         ->  Seq Scan on pg_class cl2  (cost=0.00..141.31 rows=5431 width=8) (actual time=0.007..3.292 rows=3474 loops=1)
         ->  Hash  (cost=222.93..222.93 rows=2 width=4) (actual time=7.143..7.143 rows=1 loops=1)
               Buckets: 131072  Batches: 1  Memory Usage: 1kB
               ->  Hash Join  (cost=66.13..222.93 rows=4 width=4) (actual time=6.164..7.135 rows=1 loops=1)
                     Hash Cond: (cl1.relnamespace = n1.oid)
                     Extra Text: Hash chain length 1.0 avg, 1 max, using 1 of 131072 buckets.
                     ->  Hash Join  (cost=65.03..221.79 rows=4 width=8) (actual time=6.063..6.882 rows=1 loops=1)
                           Hash Cond: (pp.oid = pr.paroid)
                           Join Filter: ((SubPlan 1) OR (('2024-04-15 14:49:49.235827+10'::timestamp with time zone >= ("substring"(pg_get_expr(pr.parrangestart, pr.parchildrelid), '#"%#"::%'::text, '#'::text))::timestamp without t
ime zone) AND ('2024-04-15 14:49:49.235827+10'::timestamp with time zone <= ("substring"(pg_get_expr(pr.parrangeend, pr.parchildrelid), '#"%#"::%'::text, '#'::text))::timestamp without time zone)))
                           Extra Text: Hash chain length 2.0 avg, 6 max, using 991 of 32768 buckets.
                           ->  Hash Join  (cost=7.01..161.94 rows=4 width=40) (actual time=0.444..1.112 rows=1 loops=1)
                                 Hash Cond: (cl1.oid = pp.parrelid)
                                 Extra Text: Hash chain length 1.0 avg, 2 max, using 992 of 65536 buckets.
                                 ->  Seq Scan on pg_class cl1  (cost=0.00..154.89 rows=1 width=8) (actual time=0.008..0.561 rows=1 loops=1)
                                       Filter: (relname = 'db_files_history'::name)
                                       Rows Removed by Filter: 3473
                                 ->  Hash  (cost=7.00..7.00 rows=1 width=40) (actual time=0.373..0.373 rows=1001 loops=1)
                                       Buckets: 65536  Batches: 1  Memory Usage: 63kB
                                       ->  Seq Scan on pg_partition pp  (cost=0.00..7.00 rows=1 width=40) (actual time=0.005..0.206 rows=1001 loops=1)
                                             Filter: ((NOT paristemplate) AND (parkind = 'r'::"char"))
                           ->  Hash  (cost=58.00..58.00 rows=1 width=72) (actual time=0.874..0.874 rows=2001 loops=1)
                                 Buckets: 32768  Batches: 1  Memory Usage: 712kB
                                 ->  Seq Scan on pg_partition_rule pr  (cost=0.00..58.00 rows=1 width=72) (actual time=0.003..0.504 rows=2001 loops=1)
                                       Filter: parrangestartincl
                                       Rows Removed by Filter: 1
                           SubPlan 1
                             ->  Materialize  (cost=0.00..3.54 rows=4 width=4) (actual time=0.023..0.023 rows=1 loops=1)
                                   ->  Seq Scan on pg_opclass  (cost=0.00..3.53 rows=4 width=4) (actual time=0.008..0.020 rows=3 loops=1)
                                         Filter: (opcname = 'timestamp_ops'::name)
                                         Rows Removed by Filter: 199
                     ->  Hash  (cost=1.09..1.09 rows=1 width=4) (actual time=0.006..0.006 rows=1 loops=1)
                           Buckets: 131072  Batches: 1  Memory Usage: 1kB
                           ->  Seq Scan on pg_namespace n1  (cost=0.00..1.09 rows=1 width=4) (actual time=0.004..0.005 rows=1 loops=1)
                                 Filter: (nspname = 'arenadata_toolkit'::name)
                                 Rows Removed by Filter: 7
   ->  Hash  (cost=1.09..1.09 rows=1 width=4) (actual time=0.018..0.018 rows=1 loops=1)
         Buckets: 131072  Batches: 1  Memory Usage: 1kB
         ->  Seq Scan on pg_namespace n2  (cost=0.00..1.09 rows=1 width=4) (actual time=0.015..0.016 rows=1 loops=1)
               Filter: (nspname = 'arenadata_toolkit'::name)
               Rows Removed by Filter: 7
 Planning time: 5.906 ms
   (slice0)    Executor memory: 16689K bytes.  Work_mem: 712K bytes max.
 Memory used:  128000kB
 Optimizer: Postgres query optimizer
 Execution time: 12.830 ms
(51 rows)
```
<details>

Reply:
and a bit different option  (with eliminated subquery):

<details>

<summary> log </summary>

```
postgres=# 
postgres=# explain analyze SELECT
                FROM pg_partition pp
                JOIN pg_class cl1 ON pp.parrelid = cl1.oid
                JOIN pg_partition_rule pr ON pr.paroid = pp.oid
                JOIN pg_class cl2 ON cl2.oid = pr.parchildrelid
                JOIN pg_namespace n1 ON cl1.relnamespace = n1.oid
                JOIN pg_namespace n2 ON cl2.relnamespace = n2.oid
                WHERE
                        CASE
                                WHEN pp.parclass[0] = ANY('{3128,10052,10155}'::oid[])
                                THEN now() BETWEEN
                                        CAST(substring(pg_get_expr(pr.parrangestart, pr.parchildrelid) FROM '#"%#"::%' FOR '#')
                                                AS TIMESTAMP WITHOUT TIME ZONE)
                                        AND
                                        CAST(substring(pg_get_expr(pr.parrangeend, pr.parchildrelid) FROM '#"%#"::%' FOR '#')
                                                AS TIMESTAMP WITHOUT TIME ZONE)
                                ELSE FALSE
                        END
                        AND
                        cl1.relname = 'db_files_history'
                        AND
                        n1.nspname = 'arenadata_toolkit'
                        AND
                        n2.nspname = 'arenadata_toolkit'
                        AND
                        pp.paristemplate = false
                        AND
                        pp.parkind = 'r'  --range
                        AND
                        pr.parrangestartincl = 't';
                                                                                                                                                                                                                                       
           QUERY PLAN                                                                                                                                                                                                                  
                                 
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
---------------------------------
 Hash Join  (cost=222.31..384.07 rows=4 width=0) (actual time=6.422..7.837 rows=1 loops=1)
   Hash Cond: (cl2.relnamespace = n2.oid)
   Extra Text: Hash chain length 1.0 avg, 1 max, using 1 of 131072 buckets.
   ->  Hash Join  (cost=221.21..382.92 rows=4 width=4) (actual time=6.207..7.481 rows=1 loops=1)
         Hash Cond: (cl2.oid = pr.parchildrelid)
         Extra Text: Hash chain length 1.0 avg, 1 max, using 1 of 131072 buckets.
         ->  Seq Scan on pg_class cl2  (cost=0.00..141.31 rows=5431 width=8) (actual time=0.009..0.641 rows=3474 loops=1)
         ->  Hash  (cost=221.17..221.17 rows=2 width=4) (actual time=6.066..6.066 rows=1 loops=1)
               Buckets: 131072  Batches: 1  Memory Usage: 1kB
               ->  Hash Join  (cost=66.13..221.17 rows=4 width=4) (actual time=5.107..6.063 rows=1 loops=1)
                     Hash Cond: (cl1.relnamespace = n1.oid)
                     Extra Text: Hash chain length 1.0 avg, 1 max, using 1 of 131072 buckets.
                     ->  Hash Join  (cost=65.03..220.02 rows=4 width=8) (actual time=5.008..5.788 rows=1 loops=1)
                           Hash Cond: (pp.oid = pr.paroid)
                           Join Filter: CASE WHEN (pp.parclass[0] = ANY ('{3128,10052,10155}'::oid[])) THEN (('2024-04-15 15:02:37.046714+10'::timestamp with time zone >= ("substring"(pg_get_expr(pr.parrangestart, pr.parchildrelid), '#"%#"::%'::text, '#'::text))::timestamp without time zone) AND ('2024-04-15 15:02:37.046714+10'::timestamp with time zone <= ("substring"(pg_get_expr(pr.parrangeend, pr.parchildrelid), '#"%#"::%'::text, '#'::text))::timestamp without time zone)) ELSE false END
                           Extra Text: Hash chain length 2.0 avg, 6 max, using 991 of 32768 buckets.
                           ->  Hash Join  (cost=7.01..161.94 rows=4 width=40) (actual time=0.429..1.070 rows=1 loops=1)
                                 Hash Cond: (cl1.oid = pp.parrelid)
                                 Extra Text: Hash chain length 1.0 avg, 2 max, using 992 of 65536 buckets.
                                 ->  Seq Scan on pg_class cl1  (cost=0.00..154.89 rows=1 width=8) (actual time=0.011..0.552 rows=1 loops=1)
                                       Filter: (relname = 'db_files_history'::name)
                                       Rows Removed by Filter: 3473
                                 ->  Hash  (cost=7.00..7.00 rows=1 width=40) (actual time=0.368..0.368 rows=1001 loops=1)
                                       Buckets: 65536  Batches: 1  Memory Usage: 63kB
                                       ->  Seq Scan on pg_partition pp  (cost=0.00..7.00 rows=1 width=40) (actual time=0.004..0.188 rows=1001 loops=1)
                                             Filter: ((NOT paristemplate) AND (parkind = 'r'::"char"))
                           ->  Hash  (cost=58.00..58.00 rows=1 width=72) (actual time=0.850..0.850 rows=2001 loops=1)
                                 Buckets: 32768  Batches: 1  Memory Usage: 712kB
                                 ->  Seq Scan on pg_partition_rule pr  (cost=0.00..58.00 rows=1 width=72) (actual time=0.005..0.492 rows=2001 loops=1)
                                       Filter: parrangestartincl
                                       Rows Removed by Filter: 1
                     ->  Hash  (cost=1.09..1.09 rows=1 width=4) (actual time=0.005..0.005 rows=1 loops=1)
                           Buckets: 131072  Batches: 1  Memory Usage: 1kB
                           ->  Seq Scan on pg_namespace n1  (cost=0.00..1.09 rows=1 width=4) (actual time=0.004..0.004 rows=1 loops=1)
                                 Filter: (nspname = 'arenadata_toolkit'::name)
                                 Rows Removed by Filter: 7
   ->  Hash  (cost=1.09..1.09 rows=1 width=4) (actual time=0.016..0.016 rows=1 loops=1)
         Buckets: 131072  Batches: 1  Memory Usage: 1kB
         ->  Seq Scan on pg_namespace n2  (cost=0.00..1.09 rows=1 width=4) (actual time=0.012..0.012 rows=1 loops=1)
               Filter: (nspname = 'arenadata_toolkit'::name)
               Rows Removed by Filter: 7
 Planning time: 4.664 ms
   (slice0)    Executor memory: 16657K bytes.  Work_mem: 712K bytes max.
 Memory used:  128000kB
 Optimizer: Postgres query optimizer
 Execution time: 7.978 ms
(46 rows)

postgres=# 
postgres=# 
postgres=# 
postgres=# 
postgres=# 
postgres=# 
postgres=# 
postgres=# 
postgres=# 
postgres=# 
postgres=# 
postgres=# 
postgres=# explain analyze SELECT
                FROM pg_partition pp
                JOIN pg_class cl1 ON pp.parrelid = cl1.oid
                JOIN pg_partition_rule pr ON pr.paroid = pp.oid
                JOIN pg_class cl2 ON cl2.oid = pr.parchildrelid
                JOIN pg_namespace n1 ON cl1.relnamespace = n1.oid
                JOIN pg_namespace n2 ON cl2.relnamespace = n2.oid
                WHERE
                        pp.paristemplate = false
                        AND
                        n1.nspname = 'arenadata_toolkit'
                        AND
                        n2.nspname = 'arenadata_toolkit'
                        AND
                        cl1.relname = 'db_files_history'
                        AND
                        (pp.parclass[0] != ALL ('{3128,10052,10155}'::oid[]) OR
                        now() BETWEEN CAST(
                                substring(pg_get_expr(pr.parrangestart, pr.parchildrelid) FROM '#"%#"::%' FOR '#')
                                AS TIMESTAMP WITHOUT TIME ZONE)
                        AND
                        CAST(substring(pg_get_expr(pr.parrangeend, pr.parchildrelid) FROM '#"%#"::%' FOR '#')
                                AS TIMESTAMP WITHOUT TIME ZONE)
                        )
                        AND
                        pp.parkind = 'r'  --range
                        AND
                        pr.parrangestartincl = 't';
                                                                                                                                                                                                                                      Q
UERY PLAN                                                                                                                                                                                                                              
         
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
---------
 Hash Join  (cost=222.31..384.07 rows=4 width=0) (actual time=7.859..9.714 rows=1 loops=1)
   Hash Cond: (cl2.relnamespace = n2.oid)
   Extra Text: Hash chain length 1.0 avg, 1 max, using 1 of 131072 buckets.
   ->  Hash Join  (cost=221.21..382.92 rows=4 width=4) (actual time=7.771..9.394 rows=1 loops=1)
         Hash Cond: (cl2.oid = pr.parchildrelid)
         Extra Text: Hash chain length 1.0 avg, 1 max, using 1 of 131072 buckets.
         ->  Seq Scan on pg_class cl2  (cost=0.00..141.31 rows=5431 width=8) (actual time=0.007..0.829 rows=3474 loops=1)
         ->  Hash  (cost=221.17..221.17 rows=2 width=4) (actual time=7.662..7.662 rows=1 loops=1)
               Buckets: 131072  Batches: 1  Memory Usage: 1kB
               ->  Hash Join  (cost=66.13..221.17 rows=4 width=4) (actual time=5.364..7.659 rows=1 loops=1)
                     Hash Cond: (cl1.relnamespace = n1.oid)
                     Extra Text: Hash chain length 1.0 avg, 1 max, using 1 of 131072 buckets.
                     ->  Hash Join  (cost=65.03..220.02 rows=4 width=8) (actual time=5.271..7.004 rows=1 loops=1)
                           Hash Cond: (pp.oid = pr.paroid)
                           Join Filter: ((pp.parclass[0] <> ALL ('{3128,10052,10155}'::oid[])) OR (('2024-04-15 15:03:03.222381+10'::timestamp with time zone >= ("substring"(pg_get_expr(pr.parrangestart, pr.parchildrelid), '#"%#"::%'::text, '#'::text))::timestamp without time zone) AND ('2024-04-15 15:03:03.222381+10'::timestamp with time zone <= ("substring"(pg_get_expr(pr.parrangeend, pr.parchildrelid), '#"%#"::%'::text, '#'::text))::timestamp without time zone)))
                           Extra Text: Hash chain length 2.0 avg, 6 max, using 991 of 32768 buckets.
                           ->  Hash Join  (cost=7.01..161.94 rows=4 width=40) (actual time=0.430..2.002 rows=1 loops=1)
                                 Hash Cond: (cl1.oid = pp.parrelid)
                                 Extra Text: Hash chain length 1.0 avg, 2 max, using 992 of 65536 buckets.
                                 ->  Seq Scan on pg_class cl1  (cost=0.00..154.89 rows=1 width=8) (actual time=0.010..1.434 rows=1 loops=1)
                                       Filter: (relname = 'db_files_history'::name)
                                       Rows Removed by Filter: 3473
                                 ->  Hash  (cost=7.00..7.00 rows=1 width=40) (actual time=0.371..0.371 rows=1001 loops=1)
                                       Buckets: 65536  Batches: 1  Memory Usage: 63kB
                                       ->  Seq Scan on pg_partition pp  (cost=0.00..7.00 rows=1 width=40) (actual time=0.004..0.249 rows=1001 loops=1)
                                             Filter: ((NOT paristemplate) AND (parkind = 'r'::"char"))
                           ->  Hash  (cost=58.00..58.00 rows=1 width=72) (actual time=0.874..0.874 rows=2001 loops=1)
                                 Buckets: 32768  Batches: 1  Memory Usage: 712kB
                                 ->  Seq Scan on pg_partition_rule pr  (cost=0.00..58.00 rows=1 width=72) (actual time=0.002..0.457 rows=2001 loops=1)
                                       Filter: parrangestartincl
                                       Rows Removed by Filter: 1
                     ->  Hash  (cost=1.09..1.09 rows=1 width=4) (actual time=0.005..0.005 rows=1 loops=1)
                           Buckets: 131072  Batches: 1  Memory Usage: 1kB
                           ->  Seq Scan on pg_namespace n1  (cost=0.00..1.09 rows=1 width=4) (actual time=0.004..0.004 rows=1 loops=1)
                                 Filter: (nspname = 'arenadata_toolkit'::name)
                                 Rows Removed by Filter: 7
   ->  Hash  (cost=1.09..1.09 rows=1 width=4) (actual time=0.011..0.011 rows=1 loops=1)
         Buckets: 131072  Batches: 1  Memory Usage: 1kB
         ->  Seq Scan on pg_namespace n2  (cost=0.00..1.09 rows=1 width=4) (actual time=0.009..0.009 rows=1 loops=1)
               Filter: (nspname = 'arenadata_toolkit'::name)
               Rows Removed by Filter: 7
 Planning time: 5.604 ms
   (slice0)    Executor memory: 16657K bytes.  Work_mem: 712K bytes max.
 Memory used:  128000kB
 Optimizer: Postgres query optimizer
 Execution time: 9.870 ms
(46 rows)

postgres=# 
```
</details>